### PR TITLE
KOGITO-7890 - Use operation id as work item handler name

### PR DESCRIPTION
**Pull request by cristianonicolai**
Related to https://github.com/quarkiverse/quarkus-openapi-generator/pull/137

### CODE REVIEW

**Review by ricardozanini — APPROVED** [review]

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/GeneratedRestOperationHandler.java [hunk 1]
```diff
@@ -23,17 +23,17 @@
 
 public class GeneratedRestOperationHandler implements RestOperationHandler {
 
-    private final String className;
+    private final String workItemHandler;
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/GeneratedRestOperationHandler.java:26]: I do not think this class needs to be changed.

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/RestOperationHandlerFactory.java [hunk 2]
```diff
@@ -27,8 +27,8 @@ private RestOperationHandlerFactory() {
 
     public static RestOperationHandler get(ParserContext parserContext, WorkflowOperationId id) {
         KogitoBuildContext context = parserContext.getContext();
-        String className = ServerlessWorkflowUtils.getOpenApiClassName(id.getFileName(), id.getOperation());
-        return context.getGeneratedHandlers().contains(className) ? new GeneratedRestOperationHandler(className) : new DescriptorRestOperationHandler(parserContext, id);
+        String workItemHandlerName = ServerlessWorkflowUtils.getOpenApiWorkItemName(id.getFileName(), id.getOperation());
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/RestOperationHandlerFactory.java:30]: same here, this change is not need, I guess the intention is increase readiblity, but I do not really think it increases it

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/WorkflowCodeGenUtils.java [hunk 3]
```diff
@@ -60,11 +60,6 @@ public static Stream<Workflow> getWorkflows(Stream<Path> files) {
                 .map(Optional::get);
     }
 
-    public static String getRefHandler(GeneratedFile generatedFile) {
```
Comment (fjtirado) [inline quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/WorkflowCodeGenUtils.java:63]: This is unrelated isnt it?

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java [hunk 4]
```diff
@@ -69,7 +68,6 @@ void addWorkItemHandlers(KogitoBuildContextBuildItem contextBI, CombinedIndexBui
         for (WorkflowHandlerGenerator generator : generators) {
             for (GeneratedFile generated : generator.generateHandlerClasses(context, index)) {
                 generatedFiles.add(generated);
-                context.addGeneratedHandler(WorkflowCodeGenUtils.getRefHandler(generated));
```
Comment (fjtirado) [inline quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java:72]: This forces for every future workfllowhandler generatior to invoke context.addGeneratedHandler in the generated code. 
This is not good.

**Review by fjtirado — COMMENTED** [review]
We need to discuss the implementation. 
I was not aware user has to tweak the operationid name in the flow file, but I think we should fix in a different way.

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/openapi/WorkflowOpenApiHandlerGenerator.java [hunk 5]
```diff
@@ -77,8 +77,7 @@ protected Stream<GeneratedFile> generateHandler(KogitoBuildContext context, Anno
 
     private GeneratedFile generateHandler(KogitoBuildContext context, ClassInfo classInfo, String fileName, MethodInfo m) {
         final String packageName = context.getPackageName();
-        final String methodName = m.annotation(generatedMethod).value().asString();
```
Comment (fjtirado) [inline quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/openapi/WorkflowOpenApiHandlerGenerator.java:80]: Not needed, isnt it?
You can still use methodName both for getOpenClassName and getOpenApiWorkItemNane
This also helps to review history, I got really confused because of this change

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java (hunk 4 above)
Comment (fjtirado) [reply quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java:72]: I think we need a WorkflowGeneratedFile specialization of GenerateFile to include the handler name (which is not necessaruly the handler class)

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/rpc/WorkflowRPCHandlerGenerator.java [hunk 6]
```diff
@@ -73,6 +73,7 @@ private GeneratedFile generateHandler(KogitoBuildContext context, String service
         constructor.setBody(new BlockStmt().addStatement(new MethodCallExpr(null, "super").addArgument("enumDefault").addArgument("streamTimeout")));
         clazz.addMethod("getName", Keyword.PUBLIC).setType(parseClassOrInterfaceType(String.class.getCanonicalName()))
                 .setBody(new BlockStmt().addStatement(new ReturnStmt(new StringLiteralExpr(className))));
+        context.addGeneratedHandler(className);
```
Comment (fjtirado) [inline quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/rpc/WorkflowRPCHandlerGenerator.java:76]: This duplication between WorkflowRPCHandlerGenearteor and WorkfflowOpenHandlerGenerator I would like to avoid by adding WorkflowGeneratedFile and forcing the generator to provide a handler name (different or equal to classname)

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java [hunk 7]
```diff
@@ -191,6 +191,10 @@ public static String getOpenApiClassName(String fileName, String methodName) {
         return StringUtils.ucFirst(getValidIdentifier(removeExt(fileName.toLowerCase())) + '_' + methodName);
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java:191]: As you can verify, methodName is not really sanitized here, so once the annotation value is not sanitized, this PR is not really needed, but see my general comment

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java [hunk 8]
```diff
@@ -191,6 +191,10 @@ public static String getOpenApiClassName(String fileName, String methodName) {
...
     }
 
+    public static String getOpenApiWorkItemName(String fileName, String methodName) {
+        return removeExt(fileName) + '_' + methodName;
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java:195]: The only difference between this method and the original one is that file name portion (which is not relevant for operationid resolution, the problem reported in the JIRA) will not be sanitized, so I do not really think we need this new method

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/openapi/WorkflowOpenApiHandlerGenerator.java (hunk 5 above)
Comment (fjtirado) [reply quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/openapi/WorkflowOpenApiHandlerGenerator.java:80]: ok, there was  indeed a potential problem, if the opeationid includes a not valid java idenfitier (and since it is not sanitized, the class name will be invalid). 
According to spec, operationid is  Unique string used to identify the operation. The id MUST be unique  among all operations described in the API. The operationId value is case-sensitive.  Tools and libraries MAY use the operationId to uniquely identify an  operation, therefore, it is RECOMMENDED to follow common programming  naming conventions.

So, although not strictly prohibited by the spec, this is very unlikely to happen

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/GeneratedRestOperationHandler.java (hunk 1 above)
Comment (cristianonicolai) [reply kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/GeneratedRestOperationHandler.java:26]: @fjtirado if the goal for this variable is to define the work item handler name, I think it is a valid change for make it more readable.

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java (hunk 4 above)
Comment (cristianonicolai) [reply quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java:72]: that could be an improvement for sure, but I dont think it is a must have for the PR. I can review that in a follow up.

**Comment by cristianonicolai** [comment]
> ink that with q

@fjtirado Im aware we need to update our tests, but that will only work once we upgrade to a new openapi codegen version with the related change. I'll take care of that once it is released. I can keep this JIRA open to track that.

Regarding the impl, we do need to change things in our end. The operationId name is not a sanitized string by default, we shouldnt expect that, and there is no reason for that. We can decloupe the class name from the WIH name, which can then match the operationId.

Regarding uniqueness, operationId is true that openapi codegen enforces uniquiness in the same OpenAPI spec file, but not across multiple files, hence we need to append the file name in the WIH name to avoid conflict.

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java (hunk 7 above)
Comment (cristianonicolai) [reply kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java:191]: Non sanitized string was used as the class name, and so it would generate invalid classes. We need to make a distinction between WIH class name, which,  to be honest, is mostly irrelevant to the WIH name, which should match the operationId, as a plain string, without any manipulation from our side. So I don't see how this would work without these changes.

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/openapi/WorkflowOpenApiHandlerGenerator.java (hunk 5 above)
Comment (cristianonicolai) [reply quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/openapi/WorkflowOpenApiHandlerGenerator.java:80]: @fjtirado we dont need to rely on function name as the class name anymore. The link is based on the WIH name now, so using the method name, which is sanitized ( by openapi codegen ) version of operationId, seems fine to me.

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/GeneratedRestOperationHandler.java (hunk 1 above)
Comment (fjtirado) [reply kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/rest/GeneratedRestOperationHandler.java:26]: ok, lets use workName as variable name.

File: quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java (hunk 4 above)
Comment (fjtirado) [reply quarkus/extensions/kogito-quarkus-serverless-workflow-extension/kogito-quarkus-serverless-workflow-deployment/src/main/java/org/kie/kogito/quarkus/serverless/workflow/deployment/ServerlessWorkflowAssetsProcessor.java:72]: I disagree, the changes in the PR (which are just needed to avoid a corner case, when the unsanitized name is not a valid) introduce an issue in our design that was not there

**Comment by fjtirado** [comment]
@cristianonicolai 
I agree this change is neededd to avoid the problem with unsanitized method name. 
What I propose is to keep this JIRA open till the quarkiverse release is updated, then add the integration test and in the meantime do the change to add the new class with the workitem name.

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java (hunk 7 above)
Comment (fjtirado) [reply kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/ServerlessWorkflowUtils.java:191]: what I was trying to point out is that there are two issues: the mapping and the unsanitize, the mapping being the important one. And the mapping will work witouth this PR (when the quarkiverse one is channve), but if the unsanitize operationId is invald string (which will be a rare ocurrence), then this PR is needed. 
I also agree we need the distinciton between workItemName and className and I vote for making it explicit by the introduction of a new class that needs to be generated by every WorkflowWorItemGenerator

**Comment by cristianonicolai** [comment]
jenkins run quarkus-lts

**Review by fjtirado — APPROVED** [review]

**Comment by fjtirado** [comment]
@cristianonicolai Thanks for adding the new class, I approved, but we still need the change in the pom and the integration test (not a new test but change the opeationid in one of the existing ones)
